### PR TITLE
NO-JIRA: add a document explaining the router's threading implementat…

### PR DESCRIPTION
**Pull request by kgiusti**
…ion.

### CODE REVIEW

**Comment by ChugR** [comment]
Very nice writeup. Ship it!

One subsystem this this document does not address is the embedded python-c interface. It is protected by a global lock:
```
qd_python_lock_state_t lock_state = qd_python_lock();
```
and it is called by both the core and worker threads.

**Comment by kgiusti** [comment]
Good catch ChugR - I'll update the PR shortly.

File: docs/notes/threading-info.txt [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+Core Thread
+===========
+
+The one core thread has several responsibilities, including:
+
+ *) Managing the forwarding state
+ *) Forwarding messages across the router
+ *) Forwarding disposition and settlement state changes across the router
+ *) Managing link credit flow
+ *) Running the management Agent
```
Comment (ted-ross) [inline docs/notes/threading-info.txt]: I would rephrase this:  Responding to management requests for data owned by the core thread.

File: docs/notes/threading-info.txt [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+message and associated delivery state to the core thread.  The core
+thread uses address information from the message to determine the
+outgoing link(s) for the message.  The core then queues the message to
+the proper outgoing link(s).  The core wakes the worker thread(s) (via
+the Proton proactor) so the message(s) can be written out the Proton
+link.
+
+When delivery disposition or settlement changes are detected by a
+worker thread it notifies the core thread.  The core thread then
+finds the link or delivery to which the changes need to be propagated
```
Comment (ted-ross) [inline docs/notes/threading-info.txt]: "finds the link or delivery" suggests a search.  It's important to note that the core thread maintains linkages between incoming and outgoing deliveries and incoming and outgoing routed-links.

File: docs/notes/threading-info.txt [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+(qdr_link_t). This means that a delivery object cannot be moved from
+an inbound link to an outbound link - which seems counterintiutive to
+the process of routing a message.
+
+In reality when an inbound delivery is routed, the router create a new
+outbound delivery for each outbound link.  These outbound deliveries
+are also represented by their own pn_delivery_t and qdr_delivery_t
+instances.
+
+The core thread links the inbound and outbound qdr_deliver_t's
```
Comment (ted-ross) [inline docs/notes/threading-info.txt]: typo

File: docs/notes/threading-info.txt [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+
+ *) the delivery's disposition (local and remote)
+ *) the delivery's settlement state
+ *) a reference to the AMQP message (qd_message_t *)
+
+The qdr_delivery_t structure is defined in src/router_core/delivery.h
+
+qd_message_t
+------------
+
```
Comment (ted-ross) [inline docs/notes/threading-info.txt]: You should mention that qdrouterd does not use pn_message_t and has created qd_message_t as a replacement.  This is in contrast to qdr_delivery_t being associated with pn_delivery_t.

File: docs/notes/threading-info.txt [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+Example: Credit Grant from Core
+-------------------------------
+
+1) The core thread is in charge of managing credit.  When the time
+comes to update or drain credit the core thread will create a
+QDR_LINK_WORK_FLOW link work instance.  The core populates this with
+credit information, including drain state if necessary. The work item
+is then queued to the qdr_link_t link_work list.  The qdr_link_t is
+then added to the proper links_with_work list in the qdr_connection_t.
+The core then then wakes a worker.
```
Comment (nicob87) [inline docs/notes/threading-info.txt]: double then

**Review by ChugR — APPROVED** [review]
This basic document should be committed. It holds a wealth of detail that help the uninitiated to reason about how the router works.

Additions and corrections could be added as necessary.
